Classifier - TESS Light Curve Viewer - Axis Scales and Single-Axis Zoom/Pan

shaunanoordin opened the pull request:
## PR Overview
Package: `lib-classifier`
Closes #210 

The goal of this PR is to add the correct axis design and zoom/pan behaviour
- [x] Users will be able to only zoom and pan on the x-axis. The y-axis will remain fixed.
  - [x] Users should only be able to pan to the start or end of the Subject data. e.g. if Subject data has x-range of `[0.0, 27.8]`, we can't pan to view `x=166613.0` 
- [x] The LCV will show scales on the x & y axis.
- [ ] ~~The Zoom buttons on the Classifier toolbar should allow users to zoom in/out of the LCV.~~ Moved to #249

Other notes:
- Additional inline documentation has been (and will continue to be) added for clarification purposes.
- Where possible, the D3 calls will be broken into their own `components/LightCurveViewer/d3/` library to enable more compartmentalised testing.

### Status
~~WIP~~
Ready for review (2018.11.06)

shaunanoordin commented:
## PR Update

Features now working:
- The LCV now shows the scales of the x and y axes, and the axes reflects the pan and zoom correctly.

New question:
- Should we lock the domain of the y-axis to a specific min/max? If yes, what's the min-y and max-y?
  - This question comes after examining different Subjects on the LCV; I've seen light curves with y-domains ranging from -50 to 50, and -10 to 10. At the moment we "fit" these y-values to fit the size of the SVG container (i.e. fit the vertical space available) which looks nice visually but makes it impossible to compare different Subjects' brightness (y-axis) levels.

shaunanoordin commented:
## PR Update

Features now working:
- The LCV now allows users to pan and zoom ONLY in the x-direction
- The LCV constrains the user's view to a pre-set range - users can only browse within the bounds of 0 <= x <= 28 and -50 <= y <= 50, matching the expected domain of the TESS data

HOWEVER:
- when I demonstrated this to Nora we realise we may **need to still allow zooming in the y-direction,** since certain stars have brightness ranges much smaller than other stars, and may require a zoom-in to even see any changes in the brightness.

The intended features of the PR will be revised after I get a chance to discuss this with Nora tomorrow (Fri 2 Nov 2018), after a proper user demo to see the LCV in action.

shaunanoordin commented:
**Light Curve Viewer: To Zoom or Not To Zoom**
Decisions made with Nora:
- We're sticking to the original design
- Users can zoom and pan only in the x-direction (as original)
- Data will by default fit the width and height of the view (as original)
- Which means most volunteers (i.e. non-experts who don't pay attention to the scale of brightness) won't easily understand the difference of brightness between Subjects, but that's not a primary concern. Our main concern is making sure people can see the dips in the context of the _current light curve they're looking at_

shaunanoordin commented:
## PR Update
Done:
- Users will be able to only zoom and pan on the x-axis. The y-axis will remain fixed.
 - Users should only be able to pan to the start or end of the Subject data. e.g. if Subject data has x-range of [0.0, 27.8], we can't pan to view x=166613.0 (see above post for re-affirmation of this design)
- The LCV will show scales on the x & y axis.

Still on the TODO:
- The Zoom buttons on the Classifier toolbar should allow users to zoom in/out of the LCV.
- new: code cleanup; many of these bits can be broken into D3 calls

shaunanoordin commented:
After a discussion with @rogerhutchings , the Zoom functionality will be moved to another PR as it will involve a fair amount of work updating the Subject Viewer store, and it'd be easier to grok if those changes were in their own PR.

shaunanoordin commented:
## PR Update
This PR has been updated with some minor cleanup.

### Status
Ready for review 👌

shaunanoordin commented:
This PR has been updated to merge the latest changes from #249.

rogerhutchings commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/LightCurveViewer/LightCurveViewer.js:
```diff
@@ -83,30 +137,93 @@ class LightCurveViewer extends Component {
+  /*
+  Initialises the D3 scatterplot chart.
+  The chart is divided into multiple layers (both functional and decorative).
+  IMPORTANT: note the order these layers are added.
```
This comment could be made more helpful by adding a note to the effect that layers are added in z-index order, lowest first

rogerhutchings commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/LightCurveViewer/LightCurveViewer.js:
```diff
@@ -40,37 +78,53 @@ class LightCurveViewer extends Component {
     if (shouldAnimate) {
       points.enter()
-        .append('circle')
+        .append('circle')  // Note: all circles are of class '.data-point'
```
If that's the case, we should probably select on that class, otherwise there's no point having it at the moment as styling is done directly on the node. Also, that would probably be a good thing to optimise with a styled component.

shaunanoordin replied:
👌 Make sense. At the moment, the only place that select `<circle>`s is in
```
  clearChart () {
    this.d3dataLayer.selectAll('circle')
      .remove()
  }
``` 
and that's pretty straightforward for me to update. Gimme a sec.

As for using styled components, do you mean instead of...
```
points.enter()
  .append('circle')  // Note: all circles are of class '.data-point'
  .call(setPointStyle)
  ...
```
we do something like...?
```
const DataPoint = styled.circle`
  fill: ${props => (props.darkTheme) ? '#fff' : '#000'};
`

points.enter()
  .append(DataPoint)
```

I haven't tried combining `styled-components` with D3 yet, but that will be worth trying out in the optimisation phase.

shaunanoordin commented:
Thanks for the review, @rogerhutchings ! I've updated the layer order comments and usage of `.data-point` vs `circle` as per your feedback.

Also:
1. I've added labels to the axis in #261, though that design needs some feedback for, ah, refinement. (Look, flipping the axes around _seemed like a really good idea on Friday afternoon,_ okay?) 
2. I'll need to update the visual design to match @beckyrother's graph design; at the moment everything's hardcoded into "light mode" and doesn't have the hooks to switch themes yet.

### Status
I'm merging this and then rebasing/updating PRs #249 (Zoom Buttons) and #261 (Axis labels, etc) 

Let's gooo!
